utils/logalloc.cc

```diff
@@ -2259,7 +2259,36 @@ allocating_section::guard::~guard() {
     shard_segment_pool.set_emergency_reserve_max(_prev);
 }
 
+void allocating_section::maybe_decay_reserve() {
+    // The decay rate is inversely proportional to the reserve
+    // (every (s_segments_per_decay/_lsa_reserve) allocations).
+    //
+    // If the reserve is high, it is expensive since we may need to 
+    // evict a lot of memory to satisfy the reserve. Hence, we are
+    // willing to risk a more frequent bad_alloc in order to decay it.
+    // The cost of a bad_alloc is also lower compared to maintaining
+    // the reserve.
+    //
+    // If the reserve is low, it is not expensive to maintain, so we
+    // decay it at a lower rate.
+
+    _remaining_lsa_segments_until_decay -= _lsa_reserve;
+    if (_remaining_lsa_segments_until_decay < 0) {
+        _remaining_lsa_segments_until_decay = s_segments_per_decay;
+        _lsa_reserve = std::max(s_min_lsa_reserve, _lsa_reserve / 2);
+        llogger.debug("Decaying LSA reserve in section {} to {} segments", this, _lsa_reserve);
+    }
+
+    _remaining_std_bytes_until_decay -= _std_reserve;
+    if (_remaining_std_bytes_until_decay < 0) {
+        _remaining_std_bytes_until_decay = s_segments_per_decay;
+        _std_reserve = std::max(s_min_std_reserve, _std_reserve / 2);
+        llogger.debug("Decaying standard allocator head-room in section {} to {} [B]", this, _std_reserve);
+    }
+}
+
 void allocating_section::reserve() {
+  maybe_decay_reserve();
```

Comment:
with_reclaiming_disabled() calls reserve() when retrying, in which case we should not attempt to decay, otherwise they may cancel each other and in the extreme case we'll not progress.

I suggest to call it once per with_reclaiming_disabled.

Reply:
You mean, if the number of attempts is equal to the number of times until decay? I think it's unlikely, but will change to reduce the risk of confusion.